[EVNT-358] Splunk Automation Setup Page

vkrizan opened the pull request:
Creates a new page for automated Splunk integration set up.

The automation:
* creates new Integration
* creates new Behavior Group for `rhel` bundle
* Links the integration and behavior group (by adding it as an action)
* Adds an action to selected event types within rhel bundle (Insights)

![Screenshot 2022-04-12 at 19-36-07 Integrations Settings](https://user-images.githubusercontent.com/7695766/163021610-0ca04f30-60f5-4c78-9259-d4f79fff0c35.png)

Besides that adds and/or refactors services:
* creates `useGetBundleByName`
* makes `bundleName` of `BehaviorGroup` optional
* creates `useGetAllEventTypes`
* create `useGetAnyBehaviorGroupByNotification`

Co-authored by @jpramos123
Replaces #264

vkrizan committed:
feat(splunk): create Splunk Automated Setup page

EVNT-358

vkrizan committed:
feat(services): create useGetBundleByName

vkrizan committed:
refactor(types): optional bundleName for BehaviorGroup

vkrizan committed:
feat(splunk): add automated steps for Splunk Setup

Creates integration, behavior group and links them together.

EVNT-358

josejulio commented on src/pages/Integrations/SplunkSetup/useSplunkSetup.tsx:
```diff
@@ -0,0 +1,162 @@
+        onProgress(`  ${behaviorGroupName} behavior group created.`);
+
+        onProgress('Adding Splunk integration as an action behavior group...');
+        updateSplunkBehaviorActions(behaviorGroup, integration);
```
I suppose the `await`  is missing here

vkrizan replied:
Thanks, added that

vkrizan committed:
feat(services): create useGetAllEventTypes

vkrizan committed:
feat(services): create useGetAnyBehaviorGroupByNotification

This can be used to query behavior groups of any type of event type.
... parametrized querying.

vkrizan commented:
@josejulio @jpramos123 ready for review. Note that we'll update the visuals (wizard) separately.

@josejulio Let me know if you'd need to have the services improvements separately. Also would it be possible to not squash it but merge as is?

Thank you.

I'll go through the lint errors tomorrow.

josejulio commented:
> @josejulio Let me know if you'd need to have the services improvements separately. Also would it be possible to not squash it but merge as is?

We usually `squash and merge` - but I saw that you took the time to split each step in a separate commit. It's fine by me to merge without squash

josejulio commented on src/pages/Integrations/SplunkSetup/useSplunkSetup.tsx:
```diff
@@ -0,0 +1,207 @@
+            throw errorObject;
+        }
+
+        return payload?.value as Integration;
```
The payload implements discrimination by `payload?.status` or `payload?.type`.
While is not required to modify this PR, one should be able to do:
```suggestion
        const { payload } = await mutate(newIntegration);
        if (payload?.status !== 200) {
            throw new Error('Invalid response');
        }

        return payload.value;
```

vkrizan replied:
I'm trusting the library to produce `errorObject` if there is an error. Comparing a single HTTP status codes is not ideal either, as it might be any of the `2XX` responses. In this POST case it should be `201 Created`. Check https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#successful_responses.

I acknowledge that even the check of errorObject does not cover all cases, i.e. the case when the payload is not correct or passed.

vkrizan replied:
Also check https://marcin-piela.github.io/react-fetching-library/#/?id=error-boundaries

vkrizan replied:
Okay, I've added additional checks for `error`. Those would cover it, when the error object is not thrown. `error` is set for non `2xx` responses.

https://github.com/marcin-piela/react-fetching-library/blob/9b0abd81cf7494931ca144a1d64952f6e0d77b18/src/client/client.ts#L77
https://developer.mozilla.org/en-US/docs/Web/API/Response/ok

josejulio replied:
One of these layers that were added is a validation layer - all the requests are validated (runtime) and discriminated types are created (compile time) to ensure the types are correct. 

It's maybe not ideal, but this allows us to detect api changes at compile time (given the openapi is updated).

Aside the `status` a type is added to the payload, in this case, if the type is `"Integration"` the payload.value is constrained to `Integration` and below cast is not needed anymore.

The implementation of these mechanisms needs to be documented and needs polishing to make it usable and friendly to the implementer, but it's something that is already there. 

![Screenshot_2022-04-13_06-50-25](https://user-images.githubusercontent.com/3845764/163184119-da687ac9-a70b-48f5-9de2-8d4f2053b42f.png)

An example on other request, the IDE will even help with hinting the possible types:

![Screenshot_2022-04-13_06-57-10](https://user-images.githubusercontent.com/3845764/163185281-f8fcf23e-259a-43d3-b854-0af2a6babf38.png)

![Screenshot_2022-04-13_06-58-51](https://user-images.githubusercontent.com/3845764/163185381-7c0b1641-361c-4fe7-a733-f291bd602edc.png)

I'm not requiring these changes on this PR. Current checks are OK  and can be refined later.

vkrizan replied:
Yeah, thanks. It would be great to have this generalized and avoid repeating it for each new api hook.

I've added couple of these notes to https://issues.redhat.com/browse/NOTIF-602

josejulio commented:
> I'll go through the lint errors tomorrow.

You can start with `yarn lint:ts:fix`

josejulio reviewed (approved):
lgtm
